File: Roadmap/29 - SOLID ISP/javascript/Chrisdev00.js
/*
 * EJERCICIO:
 * Explora el "Principio SOLID de Segregación de Interfaces (Interface Segregation Principle, ISP)" 
 * y crea un ejemplo simple donde se muestre su funcionamiento de forma correcta e incorrecta.
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un gestor de impresoras.
 * Requisitos:
 * 1. Algunas impresoras sólo imprimen en blanco y negro.
 * 2. Otras sólo a color.
 * 3. Otras son multifunción, pueden imprimir, escanear y enviar fax.
 * Instrucciones:
 * 1. Implementa el sistema, con los diferentes tipos de impresoras y funciones.
 * 2. Aplica el ISP a la implementación.
 * 3. Desarrolla un código que compruebe que se cumple el principio.
 */

// Forma incorrecta de aplicar el principio ISP

class Vehicle {
    startEngine() {
        throw new Error("Method 'startEngine()' must be implemented.");
    }  
    stopEngine() {
        throw new Error("Method 'stopEngine()' must be implemented.");
    }  
    fly() {
        throw new Error("Method 'fly()' must be implemented.");
    }  
    drive() {
        throw new Error("Method 'drive()' must be implemented.");
    }  
    sail() {
        throw new Error("Method 'sail()' must be implemented.");
    }
}

class Car extends Vehicle{
    startEngine() {
        console.log('Car engine started');
    }
    
    stopEngine() {
        console.log('Car engine stopped');
    }
    
    fly() {
        throw new Error('Cars cannot fly');
    }
    
    drive() {
        console.log('Car is driving');
    }
    
    sail() {
        throw new Error('Cars cannot sail');
    }
}

const car1 = new Car();
car1.startEngine();
car1.fly();

// Forma correcta de aplicar el principio ISP

class EnginePowered{
    startEngine(){
        throw new Error("Method 'startEngine()' must be implemented.");
    }
    stopEngine(){
        throw new Error("Method 'stopEngine()' must be implemented.");
    }
}
class Drivable{
    drive(){
        throw new Error("Method 'drive()' must be implemented.");
    }
}
class Flyable{
    fly(){
        throw new Error("Method 'fly()' must be implemented.");
    }
}
class Sailable{
    sail(){
        throw new Error("Method 'sail()' must be implemented.");
    }
}


class CarEngine extends EnginePowered{
    startEngine() {
        console.log('Car engine started');
    }    
    stopEngine() {
        console.log('Car engine stopped');
    }
}

class CarDrive extends Drivable{
    drive(){
        console.log('Car is driving');
    }
}

class Car{
    constructor(){
        this.engine = new CarEngine();
        this.driving = new CarDrive();
    }
    startEngine(){
        this.engine.startEngine();
    }
    stopEngine(){
        this.engine.stopEngine();
    }
    drive(){
        this.driving.drive();
    }
}

class PlaneEngine extends EnginePowered{
    startEngine(){
        console.log('Plane engine started');
    }
    stopEngine(){
        console.log('Plane engine stopped');
    }
}

class PlaneFly extends Flyable{
    fly(){
        console.log("Plane is flying")
    }
}

class Plane{
    constructor(){
        this.engine = new PlaneEngine();
        this.flying = new PlaneFly();
    }
    startEngine(){
        this.engine.startEngine();
    }
    stopEngine(){
        this.engine.stopEngine();
    }
    fly(){
        this.flying.fly();
    }
}


const car = new Car();

car.startEngine();
car.drive();
car.stopEngine();

const plane = new Plane();

plane.startEngine();
plane.fly();
plane.stopEngine();


//////////////// --------------------------------------- EXTRA ------------------------------- //////////////////////


class BlackWhitePrint{
    printBlWh(){
        throw new Error("Method 'printBlWh()' must be implemented.");
    }
}
class ColorPrint{
    printColor(){
        throw new Error("Method 'printColor()' must be implemented.");
    }
}
class ScannerPrint{
    scan(){
        throw new Error("Method 'scan()' must be implemented.");
    }
}
class FaxPrint{
    sendFax(dest){
        throw new Error("Method 'sendFax()' must be implemented.");
    }
}

// Implementaciones para una impresoar Epson
class EpsonBlackWhite extends BlackWhitePrint{
    printBlWh(){
        console.log("Impresion en blanco y negro")
    }
}
class EpsonColor extends ColorPrint{
    printColor(){
        console.log("Impresion a colores")
    }
}
class EpsonScanner extends ScannerPrint{
    scan(){
        console.log("Escaneo disponible")
    }
}

class Epson{
    constructor(){
        this.blackWhite = new EpsonBlackWhite();
        this.colores = new EpsonColor();
        this.escaner = new EpsonScanner();
    }
    printBlWh(){
        this.blackWhite.printBlWh();
    }
    printColor(){
        this.colores.printColor();
    }
    scan(){
        this.escaner.scan();
    }
}

// Implementaciones para una impresora Cannon
class CannonBlackWhite extends BlackWhitePrint{
    printBlWh(){
        console.log("Impresion en blanco y negro")
    }
}
class CannonFax extends FaxPrint{
    sendFax(dest){
        console.log(`Enviando documento a ${dest}`)
    }
}

class Cannon{
    constructor(){
        this.blackWhite = new CannonBlackWhite();
        this.sendTo = new CannonFax();
    }
    printBlWh(){
        this.blackWhite.printBlWh();
    }
    sendFax(dest){
        this.sendTo.sendFax(dest);
    }
}

const epson = new Epson();
epson.printBlWh();
epson.printColor(),
epson.scan();

const cannon = new Cannon();
cannon.printBlWh();
cannon.sendFax("juan4512");

File: Roadmap/29 - SOLID ISP/python/Chrisdev00.py
"""
* EJERCICIO:
* Explora el "Principio SOLID de Segregación de Interfaces (Interface Segregation Principle, ISP)" 
* y crea un ejemplo simple donde se muestre su funcionamiento de forma correcta e incorrecta.
*
* DIFICULTAD EXTRA (opcional):
* Crea un gestor de impresoras.
* Requisitos:
* 1. Algunas impresoras sólo imprimen en blanco y negro.
* 2. Otras sólo a color.
* 3. Otras son multifunción, pueden imprimir, escanear y enviar fax.
* Instrucciones:
* 1. Implementa el sistema, con los diferentes tipos de impresoras y funciones.
* 2. Aplica el ISP a la implementación.
* 3. Desarrolla un código que compruebe que se cumple el principio.
"""

# Forma Incorrecta sin aplicar el principio ISP

from abc import ABC, abstractmethod

class Document(ABC):

    @abstractmethod
    def open (self):
        pass
    @abstractmethod
    def save (self):
        pass
    @abstractmethod
    def print (self):
        pass
    @abstractmethod
    def convert_to (self, format):
        pass

class PDFDocument(Document):

    def open(self):
        print("Opening PDF document")

    def save(self):
        print("Saving PDF document")

    def print(self):
        print("Printing PDF document")

    def convert_to(self, format):
        print(f"Converting PDF document to: {format}")

class WordDocument(Document):

    def open(self):
        print("Opening Word document")

    def save(self):
        print("Saving Word document")

    def print(self):
        print("Printing Word document")

    def convert_to(self, format):
        print(f"Converting Word document to: {format}")

class SpreadsheetDocument(Document):

    def open(self):
        print("Opening Spreadsheet document")

    def save(self):
        print("Saving Spreadsheet document")

    def print(self):
        # Hojas de calculo no necesitan imprimir
        pass

    def convert_to(self, format):
        # Hojas de calculo no necesitan convertir
        pass

pdf = PDFDocument()
pdf.open()
pdf.print()
pdf.convert_to('Word')

spreadsheet = SpreadsheetDocument()
spreadsheet.open()
spreadsheet.print() # No hace nada

# Forma Correcta de aplicar el principio ISP

class Document(ABC):
    
    @abstractmethod
    def open (self):
        pass
    @abstractmethod
    def save (self):
        pass    

class PrintDocument(Document):
    @abstractmethod
    def print(self):
        pass

class ConvertTo(Document):
    @abstractmethod
    def convert_to(self, format):
        pass

class PDFDocument(PrintDocument, ConvertTo):

    def open(self):
        print("Opening PDF document")

    def save(self):
        print("Saving PDF document")

    def print(self):
        print("Printing PDF document")

    def convert_to(self, format):
        print(f"Converting PDF document to: {format}")

class WordDocument(PrintDocument, ConvertTo):

    def open(self):
        print("Opening Word document")

    def save(self):
        print("Saving Word document")

    def print(self):
        print("Printing Word document")

    def convert_to(self, format):
        print(f"Converting Word document to: {format}")

class SpreadsheetDocument(Document):

    def open(self):
        print("Opening Spreadsheet document")

    def save(self):
        print("Saving Spreadsheet document")


pdf = PDFDocument()
pdf.open()
pdf.print()
pdf.convert_to('Word')

spreadsheet = SpreadsheetDocument()
spreadsheet.open()
spreadsheet.save()


############### --------------------------------- EXTRA --------------------------------------- ###################

class Printer(ABC):
    
    @abstractmethod
    def black_white_print(self):
        pass

class ColorPrinter(ABC):
    @abstractmethod
    def color_print(self):
        pass

class ScanerPrinter(ABC):
    @abstractmethod
    def scan(self):
        pass

class FaxPrinter(ABC):
    @abstractmethod
    def send_fax(self, dest):
        pass

class CannonPrinter(Printer, ColorPrinter, ScanerPrinter, FaxPrinter):

    def black_white_print(self):
        print("Impresion en blanco y negro")

    def color_print(self):
        print("Impresion a colores")

    def scan(self):
        print("Escaneo disponible")

    def send_fax(self, dest):
        print(f"Enviar documento a {dest}")

class EpsonPrinter(Printer, ColorPrinter, ScanerPrinter):

    def black_white_print(self):
        print("Impresion en blanco y negro")

    def color_print(self):
        print("Impresion a colores")

    def scan(self):
        print("Escaneo disponible")

class LGPrinter (Printer, FaxPrinter):

    def black_white_print(self):
        print("Impresion en blanco y negro")

    def send_fax(self, dest):
        print(f"Enviar documento a {dest}")

class ColorOnlyPrinter(ColorPrinter):

    def color_print(self):
        print("Impresión a colores")


cannon = CannonPrinter()
cannon.black_white_print()
cannon.scan()
cannon.send_fax("carlos@example.com")

epson = EpsonPrinter()
epson.color_print()
epson.scan()

lg = LGPrinter()
lg.send_fax("juan4562")

color_only = ColorOnlyPrinter()
color_only.color_print()